python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+        """Updates the name of a column.
+
+        Args:
+            path_from: The path to the column to be renamed.
+            new_name: The new path of the column.
+
+        Returns:
+            The UpdateSchema with the rename operation staged.
+        """
+        name_from = tuple(path_from.split(".")) if isinstance(path_from, str) else path_from
+        new_name = new_name.split(".")[-1]
 
-        if self._transaction is not None:
-            self._transaction._append_updates(*updates)  # pylint: disable=W0212
-            self._transaction._append_requirements(*requirements)  # pylint: disable=W0212
+        parent_path = name_from[:-1]
+        full_name_from = ".".join(name_from)
+
+        from_field = self._schema.find_field(full_name_from, self._case_sensitive)
+
+        if from_field.field_id in self._deletes:
+            raise ValueError(f"Cannot rename a column that will be deleted: {full_name_from}")
+
+        if updated := self._updates.get(from_field.field_id):
+            self._updates[from_field.field_id] = NestedField(
+                field_id=updated.field_id,
+                name=new_name,
+                field_type=updated.field_type,
+                doc=updated.doc,
+                required=updated.required,
+            )
         else:
-            table_update_response = self._table.catalog._commit_table(  # pylint: disable=W0212
-                CommitTableRequest(identifier=self._table.identifier[1:], updates=updates, requirements=requirements)
+            self._updates[from_field.field_id] = NestedField(
+                field_id=from_field.field_id,
+                name=new_name,
+                field_type=from_field.field_type,
+                doc=from_field.doc,
+                required=from_field.required,
             )
-            self._table.metadata = table_update_response.metadata
-            self._table.metadata_location = table_update_response.metadata_location
 
-    def _apply(self) -> Schema:
-        """Apply the pending changes to the original schema and returns the result.
+        if path_from in self._identifier_field_names:
+            self._identifier_field_names.remove(full_name_from)
+            self._identifier_field_names.add(f"{'.'.join(parent_path)}.{new_name}")
+
+        return self
+
+    def require_column(self, path: Union[str, Tuple[str, ...]]) -> UpdateSchema:
+        """Makes a column required.
+
+        This is a breaking change since writers have to make sure that
+        this value is not-null.
+
+        Args:
+            path: The path to the field
 
         Returns:
-            the result Schema when all pending updates are applied
+            The UpdateSchema with the requirement change staged.
         """
-        return _apply_changes(self._schema, self._adds, self._identifier_field_names)
+        self._set_column_requirement(path, True)
```

Comment:
I don't think this invocation is clear. When passing a boolean arg, can you use a named argument?

Reply:
My IDEA sets this:
<img width="592" alt="image" src="https://github.com/apache/iceberg/assets/1134248/8c0886df-3b87-436b-aa1c-2a2ca5ddc47c">

But I agree that `kwargs > args`, thanks for pointing this out